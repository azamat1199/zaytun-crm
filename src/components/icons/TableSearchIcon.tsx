import { FC } from "react";
import { IconSize } from "./icons.type";
import { twMerge } from "tailwind-merge";

interface TableSearchIconProps {
    size?: IconSize;
    className?: string;
}

/**
 * TableSearchIcon Component
 * @description A React component that renders a left-pointing arrow icon.
 * @example
 * <TableSearchIcon size="sm" className="text-red-500" />
 * @component
 * @param {Object} props - Component props.
 * @param {IconSize} [props.size='xs'] - Size of the arrow icon. Can be 'xs', 'sm', 'md', or 'lg'. Defaults to 'xs'.
 * @param {string} [props.className] - Additional CSS class names to be applied to the component.
 * @returns {JSX.Element | null} A SVG element representing the left-pointing arrow icon or null if size is not 'xs'.
 */

const TableSearchIcon: FC<TableSearchIconProps> = ({
    size = "xs",
    className,
}) => {
    const params = { className: twMerge(className) };

    switch (size) {
        case "xs":
            return (
                <svg
                    width="12"
                    height="12"
                    viewBox="0 0 12 12"
                    fill="none"
                    xmlns="http://www.w3.org/2000/svg"
                    {...params}
                >
                    <path
                        d="M10.5 10.5L8.325 8.325M9.5 5.5C9.5 7.70914 7.70914 9.5 5.5 9.5C3.29086 9.5 1.5 7.70914 1.5 5.5C1.5 3.29086 3.29086 1.5 5.5 1.5C7.70914 1.5 9.5 3.29086 9.5 5.5Z"
                        stroke="#101828"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                    />
                </svg>
            );
        case "sm":
            return (
                <svg
                    width="16"
                    height="16"
                    viewBox="0 0 16 16"
                    fill="none"
                    xmlns="http://www.w3.org/2000/svg"
                    {...params}
                >
                    <path
                        d="M14 14L11.1 11.1M12.6667 7.33333C12.6667 10.2789 10.2789 12.6667 7.33333 12.6667C4.38781 12.6667 2 10.2789 2 7.33333C2 4.38781 4.38781 2 7.33333 2C10.2789 2 12.6667 4.38781 12.6667 7.33333Z"
                        stroke="#101828"
                        stroke-width="1.2"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                    />
                </svg>
            );
        case "md":
            return (
                <svg
                    width="20"
                    height="20"
                    viewBox="0 0 20 20"
                    fill="none"
                    xmlns="http://www.w3.org/2000/svg"
                    {...params}
                >
                    <path
                        d="M17.5 17.5L13.875 13.875M15.8333 9.16667C15.8333 12.8486 12.8486 15.8333 9.16667 15.8333C5.48477 15.8333 2.5 12.8486 2.5 9.16667C2.5 5.48477 5.48477 2.5 9.16667 2.5C12.8486 2.5 15.8333 5.48477 15.8333 9.16667Z"
                        stroke="#101828"
                        stroke-width="1.4"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                    />
                </svg>
            );

        case "lg":
            return (
                <svg
                    width="24"
                    height="24"
                    viewBox="0 0 24 24"
                    fill="none"
                    xmlns="http://www.w3.org/2000/svg"
                    {...params}
                >
                    <path
                        d="M21 21L16.65 16.65M19 11C19 15.4183 15.4183 19 11 19C6.58172 19 3 15.4183 3 11C3 6.58172 6.58172 3 11 3C15.4183 3 19 6.58172 19 11Z"
                        stroke="#101828"
                        stroke-width="1.6"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                    />
                </svg>
            );

        default:
            return null;
    }
};

export default TableSearchIcon;
