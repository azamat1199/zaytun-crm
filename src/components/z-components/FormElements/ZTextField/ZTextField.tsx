import get from "lodash.get";
import { ComponentPropsWithoutRef, FC, forwardRef, ReactNode } from "react";
import { twMerge } from "tailwind-merge";
import ZFormLabel from "../ZFormLabel";
import ZHelperText from "../ZHelperText";
import Skeleton from "react-loading-skeleton";

export interface ZTextFieldProps
    extends Omit<ComponentPropsWithoutRef<"input">, "className"> {
    rootClassName?: string;
    formInputSize?: "lg" | "md";
    className?: string;
    fullWidth?: boolean;
    startIcon?: ReactNode;
    endIcon?: ReactNode;
    hasError?: boolean;
    helperText?: any;
    label?: ReactNode;
    errors?: any;
    rightElement?: ReactNode;
    startIconProps?: ComponentPropsWithoutRef<"div">;
    endIconProps?: ComponentPropsWithoutRef<"div">;
    loading?: boolean;
}

type FormInputRef = React.ComponentPropsWithRef<"input">["ref"];

/**
 * ZTextField is a customizable text input field component.
 *
 * @component
 *
 * @example
 * // Basic usage
 * <ZTextField />
 *
 * // With label, helper text, and custom icons
 * <ZTextField
 *    label="Name"
 *    helperText="Please enter your name"
 *    startIcon={<Icon />}
 *    endIcon={<Icon />}
 * />
 *
 * @param {Object} props - The props for ZTextField component.
 * @param {string|string[]} [props.rootClassName] - Additional class name(s) for the root element.
 * @param {('lg' | 'md')} [props.formInputSize='lg'] - The size of the input field.
 * @param {string|string[]} [props.className] - Additional class name(s) for the input element.
 * @param {boolean} [props.fullWidth] - Whether the input field should take full width.
 * @param {ReactNode} [props.startIcon] - Icon displayed at the start of the input field.
 * @param {ReactNode} [props.endIcon] - Icon displayed at the end of the input field.
 * @param {boolean} [props.hasError] - Indicates whether there's an error with the input.
 * @param {ReactNode} [props.helperText] - Helper text to provide additional information.
 * @param {ReactNode} [props.label] - Label for the input field.
 * @param {ReactNode} [props.rightElement] - element is rendered right side of the TextField
 * @param {Object} [props.startIconProps] - props for start icon container
 * @param {Object} [props.endIconProps] - props for end icon container
 * @param {boolean} [props.loading=false] - Whether indicate skeleton instead of TextField
 * @returns {JSX.Element} - Rendered ZTextField component.
 */

const ZTextField: FC<ZTextFieldProps> = forwardRef(
    (props, ref: FormInputRef) => {
        const {
            rootClassName,
            className = [],
            formInputSize = "lg",
            fullWidth,
            startIcon,
            endIcon,
            label,
            errors = {},
            rightElement,
            startIconProps,
            endIconProps,
            loading = false,
            ...computedProps
        } = props;

        const helperText =
            props.helperText || get(errors, `${props.name}.message`);
        const hasError =
            props.hasError || (props.name && Object.hasOwn(errors, props.name));

        return (
            <div
                className={twMerge([fullWidth && "w-full", rootClassName])}
                id={props.name || ""}
                role="group"
            >
                {label && (
                    <ZFormLabel
                        disabled={props.disabled}
                        className="mb-[6px] text-b-2-m"
                    >
                        {label}
                    </ZFormLabel>
                )}
                <div
                    className={twMerge([
                        "relative flex items-center",
                        loading && "[&>span]:w-full",
                    ])}
                >
                    {!loading ? (
                        <>
                            {startIcon && (
                                <div
                                    role="presentation"
                                    {...startIconProps}
                                    className={twMerge(
                                        "absolute left-2.5 top-2/4 -translate-y-2/4",
                                        startIconProps?.className,
                                    )}
                                >
                                    {startIcon}
                                </div>
                            )}
                            {endIcon && (
                                <div
                                    role="presentation"
                                    {...endIconProps}
                                    className={twMerge(
                                        "absolute right-2.5 top-2/4 -translate-y-2/4",
                                        endIconProps?.className,
                                    )}
                                >
                                    {endIcon}
                                </div>
                            )}
                            <input
                                ref={ref}
                                {...computedProps}
                                className={twMerge([
                                    formInputSize === "lg" && "h-12",
                                    formInputSize === "md" && "h-10",
                                    //   DEFAULT
                                    "bg-white w-full focus:outline-none focus:ring-0 transition-all duration-200 ease-in-out appearance-none outline-none rounded-md border border-c_neutral-300 px-[10px] py-4 text-b-1-r text-c_neutral-800",
                                    // DEFAULT PLACEHOLDER
                                    "placeholder:text-b-1-r  placeholder:font-normal placeholder:text-c_neutral-500",
                                    // FOCUS
                                    "focus:border-c_primary-500",
                                    // DISABLED
                                    "disabled:border-c_neutral-200 disabled:bg-c_neutral-25 disabled:cursor-not-allowed disabled:text-c_neutral-400",
                                    // READONLY
                                    "[&[readonly]]:border-c_neutral-200 [&[readonly]]:bg-c_neutral-25 [&[readonly]]:text-c_neutral-400",
                                    startIcon && "pl-10",
                                    hasError && "border-c_error-400",
                                    className,
                                ])}
                            />
                            {rightElement && rightElement}
                        </>
                    ) : (
                        <Skeleton
                            width="100%"
                            className={twMerge([
                                formInputSize === "lg" && "h-12",
                                formInputSize === "md" && "h-10",
                            ])}
                        />
                    )}
                </div>
                {helperText && (
                    <ZHelperText
                        className="mt-[6px]"
                        {...(hasError && { variant: "error" })}
                    >
                        {helperText}
                    </ZHelperText>
                )}
            </div>
        );
    },
);

ZTextField.displayName = "ZTextField";

export default ZTextField;
