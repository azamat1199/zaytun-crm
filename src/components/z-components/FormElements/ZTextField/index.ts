import ZTextField from "./ZTextField";

export default ZTextField;
