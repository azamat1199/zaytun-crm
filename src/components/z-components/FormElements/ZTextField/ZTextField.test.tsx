import { expect, test, vi, describe } from "vitest";
import { render, screen, fireEvent } from "@testing-library/react";
import ZTextField from "./ZTextField";
import TableSearchIcon from "@/components/icons/TableSearchIcon";

const handleChange = vi.fn();

const setup = () => {
    const utils = render(
        <ZTextField
            onChange={handleChange}
            placeholder="placeholder"
            helperText="helpertext"
            startIcon={<TableSearchIcon />}
            name="firstName"
            label="Something"
        />,
    );
    const input: HTMLInputElement = screen.getByPlaceholderText("placeholder");
    const inputContainer: HTMLDivElement = screen.getByRole("group");
    return {
        input,
        inputContainer,
        ...utils,
    };
};

describe("ZTextField tests", () => {
    test("Render label successfully", () => {
        render(<ZTextField label="label" />);
        expect(screen.getByText("label")).toBeDefined();
    });

    test("Render placeholder successfully", () => {
        render(<ZTextField placeholder="placeholder" />);
        expect(screen.getByPlaceholderText("placeholder")).toBeDefined();
    });

    test("Render helperText successfully", () => {
        render(<ZTextField helperText="helperText" />);
        expect(screen.getByText("helperText")).toBeDefined();
    });

    test("Onchange event works successfully", () => {
        const inputValue = "mock";
        const { input } = setup();
        fireEvent.change(input, { target: { value: inputValue } });

        expect(input.value).toBe(inputValue);
        expect(handleChange).toHaveBeenCalledTimes(1);
    });

    test("Render helper text successfully", () => {
        setup();
        const p = screen.getByText("helpertext");
        expect(p).toBeDefined();
    });

    test("Render startIcon successfully", () => {
        setup();
        const startIcon = screen.getByRole("presentation");
        expect(startIcon).toBeDefined();
    });

    test("Render endIcon successfully", () => {
        setup();
        const endIcon = screen.getByRole("presentation");
        expect(endIcon).toBeDefined();
    });

    test("renders correctly", () => {
        const { inputContainer } = setup();
        expect(inputContainer).toMatchSnapshot();
    });
});
