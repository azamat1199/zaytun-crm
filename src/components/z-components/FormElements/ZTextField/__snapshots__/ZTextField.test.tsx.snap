// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ZTextField tests > renders correctly 1`] = `
<div
  class=""
  id="firstName"
  role="group"
>
  <label
    class="font-medium mb-[6px] text-b-2-m"
  >
    Something
  </label>
  <div
    class="relative flex items-center"
  >
    <div
      class="absolute left-2.5 top-2/4 -translate-y-2/4"
      role="presentation"
    >
      <svg
        class=""
        fill="none"
        height="12"
        viewBox="0 0 12 12"
        width="12"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10.5 10.5L8.325 8.325M9.5 5.5C9.5 7.70914 7.70914 9.5 5.5 9.5C3.29086 9.5 1.5 7.70914 1.5 5.5C1.5 3.29086 3.29086 1.5 5.5 1.5C7.70914 1.5 9.5 3.29086 9.5 5.5Z"
          stroke="#101828"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </div>
    <input
      class="h-12 bg-white w-full focus:outline-none focus:ring-0 transition-all duration-200 ease-in-out appearance-none outline-none rounded-md border border-c_neutral-300 px-[10px] py-4 text-c_neutral-800 placeholder:font-normal placeholder:text-c_neutral-500 focus:border-c_primary-500 disabled:border-c_neutral-200 disabled:bg-c_neutral-25 disabled:cursor-not-allowed disabled:text-c_neutral-400 [&[readonly]]:border-c_neutral-200 [&[readonly]]:bg-c_neutral-25 [&[readonly]]:text-c_neutral-400 pl-10"
      helpertext="helpertext"
      name="firstName"
      placeholder="placeholder"
    />
  </div>
  <p
    class="text-c_neutral-500 mt-[6px]"
  >
    helpertext
  </p>
</div>
`;
